### DISCUSSION TITLE
How to Access Data from Hooks (like beforeValidate) within HasManyRelationManager

### DISCUSSION THREAD

**Discussion by intrepidws** [filamentphp/filament#1731] [Help]
I am trying to use the `beforeValidate` hook on a create form within a `HasManyRelationManager` class. While I can get the hooks to fire just fine, I cannot figure out how to actually access the data from the form. Within a `CreateRecord` class, I am able to use Using `$this->data` and it works great - but not within a `HasManyRelationManager` class. I have also tried `$this->record` but no joy.

```php
protected function beforeValidate(): void
{
    dd($this->data);
}
```

**Comment by intrepidws** [comment]
I was able to figure out that I can access the form data using `$this->mountedTableActionData['field']`, but that really only works within `beforeValidate`. Any changes made to this variable in a later hook (like `afterValidate` or `beforeCreate`) won't be picked up.

How could I make changes to the fields from within the `afterValidate` or `beforeCreate` hooks?

**Reply by danharrin** [reply]
What are you trying to do? Maybe you're looking for the `mutateFormDataBeforeSave()` method instead.

**Reply by intrepidws** [reply]
I'm trying to add some fairly complex membership system API lookup logic that, if it fails, I'd like the validation itself to fail as well.

**Reply by danharrin** [reply]
And that validation can't be tied to a specific component? You could use a custom class/closure validation rule.

After validation, you can access the form data in a hook using `$this->getMountedTableActionForm()->getState()`.

**Reply by intrepidws** [reply]
Hmm... perhaps. I'll look into that avenue, thanks.

So can you confirm that there is no method of changing the form data from within `afterValidate` or `beforeCreate`?

**Reply by danharrin** [reply]
Yeah, as I said in a previous message you can use `mutateFormDataBeforeSave()`, which is better than using hooks.

**Reply by intrepidws** [reply]
Right, understood. I use `mutateFormDataBeforeSave()` in some other places in the code, and it works well. I just wanted to confirm about `afterValidate()` and `beforeCreate`()`.

Thanks for your help.
